### PR TITLE
v0.4 axes transforms

### PR DESCRIPTION
**Pull request by will-moore**
WIP:

Test `napari https://minio-dev.openmicroscopy.org/idr/v0.4/2022-01-05/idr0062/6001240.zarr`.
To show scale-bar, open the napari terminal and (on latest napari):
```
viewer.scale_bar.visible = True
```
(This *was previously* `viewer.scale_bar_visible` https://github.com/napari/napari/pull/1720)

This PR requires https://pypi.org/project/ome-zarr/0.3a1/

To do:
 - [x] handle axes in v0.4 format
 - [x] handle transform scale to set pixel sizes 
 - [ ] affine transform ?!?

Not sure yet how to convert affine transform in the spec into what napari is expecting? https://napari.org/api/stable/napari.Viewer.html#napari.Viewer.add_image

```
- scale (tuple of float or list) – Scale factors for the layer. If a list then must be a list of tuples of float with the same length as the axis that is being expanded as channels.

- affine (n-D array or napari.utils.transforms.Affine) – (N+1, N+1) affine transformation matrix in homogeneous coordinates. The first (N, N) entries correspond to a linear transform and the final column is a length N translation vector and a 1 or a napari Affine transform object. Applied as an extra transform on top of the provided scale, rotate, and shear values.
```

### CODE REVIEW

**Comment by will-moore** [comment]
Without scale metadata, scale-bar shows pixels. 
NB: size Z is 235.

Screenshot 2022-01-18 at 12 38 52

Using pixel sizes in microns to add:
 `"scale": (0.5002025531914894, 0.3603981534640209, 0.3603981534640209)`
to the metadata fixes the scale-bar to show correct size (compare with IDR image below), but now there are 470 Z-sections and if you go above 235 on the Z-slider you just get black.

Screenshot 2022-01-18 at 12 40 34

See IDR https://idr.openmicroscopy.org/webclient/img_detail/6001240/

Screenshot 2022-01-18 at 12 42 31

**Comment by will-moore** [comment]
You can also set units with ```viewer.scale_bar.unit='um'``` for microns.

**Comment by will-moore** [comment]
If I set the scale for only the X and Y axes, then I get a correct scale-bar without breaking the Z-slider:

`"scale": (1, 0.36, 0.36)`,

but this distorts the 3D view:

Screenshot 2022-01-18 at 13 06 50

**Comment by will-moore** [comment]
Also see https://forum.image.sc/t/setting-scale-bar-units-in-other-than-pixels-real-coordinates/49158/12

**Comment by jni** [comment]
@will-moore what version of napari are you on? Do you think you would be able to make a reproducible example with just NumPy arrays? This works fine on main:

```python
import napari
import numpy as np

_ = napari.view_image(
    np.random.random((10, 256, 256)),
    scale=(0.5, 0.3, 0.3),
)
napari.run()
```

I get a slider with 10 positions from 0 to 9 inclusive. The broadcasting of the scale doesn't seem to be the issue either, as the sliders also seem fine with this:

```python
import napari
import numpy as np

_ = napari.view_image(
    np.random.random((3, 10, 256, 256)),
    scale=(0.5, 0.3, 0.3),
)
napari.run()
```

**Comment by will-moore** [comment]
@jni Thanks for that. I'm using `napari 0.4.12` and your sample helped me to track down the issue. I was only applying the scaling to the image layers and not the labels layer so the sizeZ of the scaled image in real world coordinates is only a portion of the sizeZ of the labels layer.

**Comment by will-moore** [comment]
Looks like it's not possible to handle units yet - see https://github.com/napari/napari/issues/1701

**Comment by will-moore** [comment]
Currently the tests are failing because the code now depends on https://github.com/ome/ome-zarr-py/pull/124
cc @sbesson Do we need to do a pre-release of `ome-zarr-py` first and then update to that in the requirements here?

**Comment by sbesson** [comment]
>  cc @sbesson Do we need to do a pre-release of ome-zarr-py first and then update to that in the requirements here?

Yes that's one of the driver for the pre-release. The other one being https://github.com/ome/omero-cli-zarr/pull/93 which is only passing because there are no tests.

I propose we review how close we are later today on signing off on the two open HCS ome-zarr-py and then decide on a 0.3.0 pre-release?

**Comment by jburel** [comment]
tested using ``ome-zarr==0.3a1``.
Scalebar is at ``25microns`` as expected.

File: napari_ome_zarr/_reader.py [hunk 1]
```diff
@@ -95,27 +118,31 @@ def f(*args: Any, **kwargs: Any) -> List[LayerData]:
...
-                    channel_axis = None
-                    if "axes" in node.metadata:
-                        # version 0.3 or greater. NB: is 'axes' optional?
-                        if "c" in node.metadata["axes"]:
-                            channel_axis = node.metadata["axes"].index("c")
-                    elif shape[CHANNEL_DIMENSION] > 1:
-                        # versions of ome-zarr-py before v0.3 support
-                        channel_axis = CHANNEL_DIMENSION
-
+                    LOGGER.debug("napari-ome-zarr: node.metadata: %s" % node.metadata)
```
Comment (joshmoore) [inline napari_ome_zarr/_reader.py:145]: very minor issue but shouldn't `napair-ome-zarr` be in the logger name?

File: napari_ome_zarr/_reader.py [hunk 2]
```diff
@@ -95,27 +118,31 @@ def f(*args: Any, **kwargs: Any) -> List[LayerData]:
...
-                shape = data[0].shape
 
                 layer_type: str = "image"
+                channel_axis = None
+                try:
+                    ch_types = [axis["type"] for axis in node.metadata["axes"]]
+                    if "channel" in ch_types:
+                        channel_axis = ch_types.index("channel")
+                except:
+                    LOGGER.error("Error reading axes: Please update ome-zarr")
```
Comment (joshmoore) [inline napari_ome_zarr/_reader.py:129]: Should this throw or perhaps return?

File: napari_ome_zarr/_reader.py [hunk 3]
```diff
@@ -84,6 +83,30 @@ def transform_properties(props=None):
     return properties
 
 
+def transform_scale(node_metadata, metadata, channel_axis, shape):
+    """
+    e.g. transformation is {"axisIndices": [1, 2, 3], "scale": [0.2, 0.06, 0.06]}
+    Get a list of these for each level in data. Just use first?
+    """
+    if "transformations" in node_metadata:
+        level_0_transforms = node_metadata["transformations"][0]
```
Comment (joshmoore) [inline napari_ome_zarr/_reader.py:92]: Just a reminder that here we should probably think of a proper model (even if just using `@property`s) as the mode

**Review by joshmoore — COMMENTED** [review]
Few minor points but happy to get this in for the meeting. I assume we keep it at a pre-release version until ome-zarr-py is ready.

**Comment by sbesson** [comment]
@will-moore can any of the three points raised in https://github.com/ome/napari-ome-zarr/pull/31#pullrequestreview-862111079 be addressed/resolved quickly? Otherwise, I agree to get this merged and tagged as a pre-release in preparation of the community call.

File: napari_ome_zarr/_reader.py (hunk 1 above)
Comment (will-moore) [reply napari_ome_zarr/_reader.py:145]: I don't see any indication of where logging statements are coming from.
e.g.
```
$ napari v4_omero/52.zarr/ -vvv
...
22:55:12 DEBUG returning /Users/wmoore/Desktop/ZARR/data/v4_omero/52.zarr [zgroup]
22:55:12 DEBUG transforming /Users/wmoore/Desktop/ZARR/data/v4_omero/52.zarr [zgroup]
22:55:12 DEBUG napari-ome-zarr: node.metadata: {'axes': [{'name': 'c', 'type...
```
But I can remove the `napari-ome-zarr` just to be consistent.

File: napari_ome_zarr/_reader.py (hunk 2 above)
Comment (will-moore) [reply napari_ome_zarr/_reader.py:129]: Ah, yes - this was added before we depended on `ome-zarr-py` version that always includes `axes` metadata.
I'll add a `raise` now.

Comment (will-moore) [reply napari_ome_zarr/_reader.py:129]: Fixed in b3abf92

File: napari_ome_zarr/_reader.py (hunk 1 above)
Comment (will-moore) [reply napari_ome_zarr/_reader.py]: Fixed in b3abf92

**Comment by sbesson** [comment]
Thanks @will-moore. Merging and creating a pre-release

**Pull request merged**